### PR TITLE
[UIKit] Add Catalyst support

### PR DESCRIPTION
**Pull request by chamons**
This get the "ignoring the union of other ignore files" for Catalyst UIKit to a pretty small subset (8 lines).

I'm going to work on those, and if they are trivial they'll stream in here, but this is the majority of it. 

Given the number of #if's I wanted to get this out for review earlier.

### CODE REVIEW

**Comment by chamons** [comment]
Open question - 5 of the unique lines:

```
!missing-selector! +NSToolbarItem::itemWithItemIdentifier:barButtonItem: not bound
!missing-selector! NSSharingServicePickerToolbarItem::activityItemsConfiguration not bound
!missing-selector! NSSharingServicePickerToolbarItem::setActivityItemsConfiguration: not bound
!missing-selector! NSSharingServicePickerTouchBarItem::activityItemsConfiguration not bound
!missing-selector! NSSharingServicePickerTouchBarItem::setActivityItemsConfiguration: not bound
```

don't really show up when I sharpie query them. 

Example:

```
$ sharpie query ios15.0-macabi-x86_64.pch -b -n NSSharingServicePickerToolbarItem

$ sharpie query ios15.0-macabi-x86_64.pch -b -n itemWithItemIdentifier

```

I'm going to do some digging, but if people see anything obvious I'm all ears.

**Comment by chamons** [comment]
Last question is on:

```
!missing-protocol-conformance! UIResponder should conform to NSTouchBarProvider (defined in 'NSTouchBarProvider' category)
```

I am trying to add it to UIResponder:


```
	[BaseType (typeof (NSObject))]
	interface UIResponder : UIAccessibilityAction, UIAccessibilityFocus, UIUserActivityRestoring
#if !TVOS
	, UIAccessibilityDragging
#endif // !TVOS
#if IOS
	, UIPasteConfigurationSupporting
#endif // IOS
#if __MACCATALYST__
	, NSTouchBarProvider
#endif
```

but it doesn't seem to be taking effect (locally?).

Let's see what xtro on CI says.

File: src/uikit.cs [hunk 1]
```diff
@@ -10397,6 +10397,11 @@ partial interface UIReferenceLibraryViewController : NSCoding {
 		IntPtr Constructor (string term);
 	}
 
+#if __MACCATALYST__
+	interface NSTouchBarProvider { }
+	interface NSTouchBar { }
+#endif
```
Comment (mandel-macaque) [inline src/uikit.cs:10403]: Why here? What are you trying to fix? What is the diff with:
```csharp
using NSTouchBarProvider = NSObject;
etc
```

File: src/uikit.cs [hunk 2]
```diff
@@ -10405,6 +10410,9 @@ interface UIResponder : UIAccessibilityAction, UIAccessibilityFocus, UIUserActiv
 #if IOS
 	, UIPasteConfigurationSupporting
 #endif // IOS
+#if __MACCATALYST__
+	, NSTouchBarProvider
```
Comment (mandel-macaque) [inline src/uikit.cs:10414]: Empty interface?

File: src/uikit.cs [hunk 3]
```diff
@@ -10614,6 +10622,29 @@ interface UIResponder : UIAccessibilityAction, UIAccessibilityFocus, UIUserActiv
...
+		[MacCatalyst (13, 0)]
+		[Export ("makeTouchBar")]
+		[return: NullAllowed]
+		NSTouchBar CreateTouchBar ();
+
+		[MacCatalyst (13, 0)]
+		[Export ("touchBar", ArgumentSemantic.Strong)]
+		[NullAllowed]
+		NSTouchBar TouchBar { get; set; }
+#endif
```
Comment (mandel-macaque) [inline src/uikit.cs:10647]: You can do the using to be NSObject, then use [NoXYZ] which is simpler to parse than #if. I would just use #if it they are due to some struct or manual code in not processed file.

File: src/uikit.cs [hunk 4]
```diff
@@ -15355,6 +15386,17 @@ interface UIViewController : NSCoding, UIAppearanceContainer, UIContentContainer
 		[NoWatch, NoTV, iOS (14,0)]
 		[Export ("setNeedsUpdateOfPrefersPointerLocked")]
 		void SetNeedsUpdateOfPrefersPointerLocked ();
+
+		[NoiOS][NoTV][NoWatch]
+		[MacCatalyst (13,0)]
```
Comment (mandel-macaque) [inline src/uikit.cs:15391]: MacCatalyst attr is not needed since min is 13.1

File: src/uikit.cs [hunk 5]
```diff
@@ -15355,6 +15386,17 @@ interface UIViewController : NSCoding, UIAppearanceContainer, UIContentContainer
...
+		[Export ("setNeedsTouchBarUpdate")]
+		void SetNeedsTouchBarUpdate ();
+
+		[NoiOS][NoTV][NoWatch]
+		[MacCatalyst (13, 0)]
```
Comment (mandel-macaque) [inline src/uikit.cs:15396]: same.

File: src/uikit.cs [hunk 6]
```diff
@@ -20994,6 +21036,14 @@ interface UIWindowScene {
...
 		[NullAllowed, Export ("focusSystem")]
 		UIFocusSystem FocusSystem { get; }
+
+#if __MACCATALYST__
+		[NoWatch][NoTV][NoiOS]
+		[MacCatalyst (13, 0)]
+		[Export ("titlebar")]
+		[NullAllowed]
+		UITitlebar Titlebar { get; }
+#endif
```
Comment (mandel-macaque) [inline src/uikit.cs:21046]: Same, no need for the attr. Also the #if is probably something we can remove.

File: src/uikit.cs [hunk 7]
```diff
@@ -23146,4 +23196,53 @@ interface UIPointerAccessory : NSCopying {
 		UIPointerAccessory CreateArrow (UIPointerAccessoryPosition position);
 	}
 
+	[MacCatalyst (13,0)][NoiOS][NoTV][NoWatch]
+	[Native]
+	public enum UITitlebarTitleVisibility : long
+	{
+		Visible,
+		Hidden
```
Comment (mandel-macaque) [inline src/uikit.cs:23204]: missing ,

File: src/uikit.cs [hunk 8]
```diff
@@ -23146,4 +23196,53 @@ interface UIPointerAccessory : NSCopying {
...
+
+	[MacCatalyst (14,0)][NoiOS][NoTV][NoWatch]
+	[Native]
+	public enum UITitlebarToolbarStyle : long
+	{
+		Automatic,
+		Expanded,
+		Preference,
+		Unified,
+		UnifiedCompact
```
Comment (mandel-macaque) [inline src/uikit.cs:23215]: missing ,

File: src/uikit.cs [hunk 9]
```diff
@@ -23146,4 +23196,53 @@ interface UIPointerAccessory : NSCopying {
...
+	}
+
+#if __MACCATALYST__
+	interface NSToolbar {}
+#endif
```
Comment (mandel-macaque) [inline src/uikit.cs:23220]: Same comment. Why not NSObject.

File: src/uikit.cs [hunk 10]
```diff
@@ -23146,4 +23196,53 @@ interface UIPointerAccessory : NSCopying {
...
+
+	[MacCatalyst (13,0)][NoiOS][NoTV][NoWatch]
```
Comment (mandel-macaque) [inline src/uikit.cs:23222]: No need for MacCatalyst.

File: src/uikit.cs [hunk 11]
```diff
@@ -23146,4 +23196,53 @@ interface UIPointerAccessory : NSCopying {
...
+		UITitlebarToolbarStyle ToolbarStyle { get; set; }
+
+		[MacCatalyst (14, 0)]
+		[Export ("separatorStyle", ArgumentSemantic.Assign)]
+		UITitlebarSeparatorStyle SeparatorStyle { get; set; }
+
+#if __MACCATALYST__
+		[NullAllowed, Export ("toolbar", ArgumentSemantic.Strong)]
+		NSToolbar Toolbar { get; set; }
+#endif
```
Comment (mandel-macaque) [inline src/uikit.cs:23240]: The #if can be remove in favour of NoXY

File: src/appkit.cs [hunk 12]
```diff
@@ -19583,6 +19595,14 @@ interface NSToolbarItem : NSCopying, NSMenuItemValidation, NSValidatedUserInterf
 		[Mac (11, 0), iOS (14, 0)]
 		[Export ("navigational")]
 		bool Navigational { [Bind ("isNavigational")] get; set; }
+
+#if __MACCATALYST__
+		[NoMac]
+		[MacCatalyst (13, 0)]
```
Comment (mandel-macaque) [inline src/appkit.cs:19601]: No need for the MacCatalyst attr. No need for the #if and if we use an #if we do not need the NoMac.

File: src/uikit.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/uikit.cs:10403]: Else, even using AppKit will solve it.

**Review by mandel-macaque — CHANGES_REQUESTED** [review]
There is a better approach. 
```
#If __MACCATALUST__
using AppKit;
#else
using NSTouchBarProvider = NSObject;
using NSTouchBar = NSObject;
#endif
```

With the above you remove ALL those new #if that make it hard to parse the code.

File: src/appkit.cs [hunk 13]
```diff
@@ -19501,6 +19509,10 @@ interface NSObject_NSToolbarItemValidation
 		bool ValidateToolbarItem (NSToolbarItem item);
 	}
 
+#if __MACCATALYST__
+	interface UIMenuElement {}
```
Comment (rolfbjarne) [inline src/appkit.cs:19513]: This should be defined already in `UIKit`

**Review by rolfbjarne — CHANGES_REQUESTED** [review]
And Manuel's comments

**Comment by mandel-macaque** [comment]
@chamons i know why it does not take effect (the interface) __MACCATALYAT__ ifs have to be BEFORE iOS, else the iOS one takes effect

File: src/appkit.cs (hunk 12 above)
Comment (chamons) [reply src/appkit.cs:19600]: I tried removing this, and we started generating the bit completely on catalyst.

**Comment by chamons** [comment]
The "unnecessary" MacCatalysts are required due to https://github.com/xamarin/xamarin-macios/issues/12824

File: src/appkit.cs (hunk 12 above)
Comment (mandel-macaque) [reply src/appkit.cs:19600]: I created an issue for this: https://github.com/xamarin/xamarin-macios/issues/12824

File: src/uikit.cs (hunk 4 above)
Comment (mandel-macaque) [reply src/uikit.cs:15394]: Ignore due to https://github.com/xamarin/xamarin-macios/issues/12824

File: src/uikit.cs (hunk 5 above)
Comment (mandel-macaque) [reply src/uikit.cs:15399]: Ignore due to https://github.com/xamarin/xamarin-macios/issues/12824

File: src/uikit.cs [hunk 14]
```diff
@@ -23146,4 +23197,47 @@ interface UIPointerAccessory : NSCopying {
...
+	{
+		Automatic,
+		Expanded,
+		Preference,
+		Unified,
+		UnifiedCompact,
+	}
+
+	[NoiOS][NoTV][NoWatch][MacCatalyst (13,0)]
+	[BaseType (typeof(NSObject))]
```
Comment (mandel-macaque) [inline src/uikit.cs:23220]:
```suggestion
	[BaseType (typeof (NSObject))]
```

File: src/appkit.cs [hunk 15]
```diff
@@ -26982,15 +27000,20 @@ interface NSSwitch : NSAccessibilitySwitch
 	}
 
 	[Mac (10,15)]
-	[NoMacCatalyst]
+	[MacCatalyst(13, 0)]
```
Comment (mandel-macaque) [inline src/appkit.cs:27003]:
```suggestion
	[MacCatalyst (13, 0)]
```

**Review by mandel-macaque — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

**Review by rolfbjarne — APPROVED** [review]

**Comment by chamons** [comment]
Legit test failure:

```
Selector not found for AppKit.NSMenuToolbarItem : itemMenu in Foundation.NSObject get_ItemMenu() on AppKit.NSMenuToolbarItem
Selector not found for AppKit.NSMenuToolbarItem : setItemMenu: in Void set_ItemMenu(Foundation.NSObject) on AppKit.NSMenuToolbarItem
```

**Review by tj-devel709 — APPROVED** [review]

**Comment by chamons** [comment]
Ok, introspection passes now. 

@mandel-macaque - Can you test this locally on beta OS?

**Comment by chamons** [comment]
Test Failures 🤔  :

```
Cecil.Tests.MarshalAsTest.TestAssembly("/Users/builder/azdo/_work/1/s/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/32bits/iOS/Xamarin.iOS.dll"):
Methods with bool return type / parameters and no MarshalAs attribute.
Expected: <empty>
But was: < "Found 1 issues with System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint):
    The char return type does not have a MarshalAs attribute: System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint)" >
Cecil.Tests.MarshalAsTest.TestAssembly("/Users/builder/azdo/_work/1/s/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/64bits/iOS/Xamarin.iOS.dll"):
Methods with bool return type / parameters and no MarshalAs attribute.
Expected: <empty>
But was: < "Found 1 issues with System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint):
    The char return type does not have a MarshalAs attribute: System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint)" >
Cecil.Tests.MarshalAsTest.TestAssembly("/Users/builder/azdo/_work/1/s/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/32bits/watchOS/Xamarin.WatchOS.dll"):
Methods with bool return type / parameters and no MarshalAs attribute.
Expected: <empty>
But was: < "Found 1 issues with System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint):
    The char return type does not have a MarshalAs attribute: System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint)" >
Cecil.Tests.MarshalAsTest.TestAssembly("/Users/builder/azdo/_work/1/s/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/64bits/tvOS/Xamarin.TVOS.dll"):
Methods with bool return type / parameters and no MarshalAs attribute.
Expected: <empty>
But was: < "Found 1 issues with System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint):
    The char return type does not have a MarshalAs attribute: System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint)" >
Cecil.Tests.MarshalAsTest.TestAssembly("/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/Xamarin.Mac.dll"):
Methods with bool return type / parameters and no MarshalAs attribute.
Expected: <empty>
But was: < "Found 1 issues with System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint):
    The char return type does not have a MarshalAs attribute: System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint)" >
Cecil.Tests.MarshalAsTest.TestAssembly("/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/4.5/Xamarin.Mac.dll"):
Methods with bool return type / parameters and no MarshalAs attribute.
Expected: <empty>
But was: < "Found 1 issues with System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint):
    The char return type does not have a MarshalAs attribute: System.Char CoreFoundation.CFString::CFStringGetCharacterAtIndex(System.IntPtr,System.nint)" >
```

**Comment by chamons** [comment]
Test failure is https://github.com/xamarin/maccore/issues/2519

**Comment by chamons** [comment]
On beta OS tests run by Manuel show two errors:

image

Looking at fixing those now.

**Comment by chamons** [comment]
This is paused until i install the best on my backup machine and test some things. Those constants are live in ObjC, but are part of a private UIKitCore framework.

I am uncertain how well our bindings will handle this, and need to test.

**Review by dalexsoto — APPROVED** [review]
One more

**Comment by chamons** [comment]
Hmm, both pass locally for me. The first one is confirmed network flake.

**Comment by chamons** [comment]
One of the CoreFoundation patches landed here by mistake. Fixing.

**Comment by chamons** [comment]
Test failure was 17793 signaling again.
